### PR TITLE
Add support for key rotation

### PR DESCRIPTION
**Pull request by drgora**
This PR adds a new `sc_prev_wcert_hash` optional field to parameter generation and proof creation.

This field is a requirement to implement the key rotation feature, and is set by `zend` to verify proofs.

### CODE REVIEW

File: src/mc_test_circuits/cert.rs [hunk 1]
```diff
@@ -71,6 +86,23 @@ fn enforce_cert_inputs_gadget<CS: ConstraintSystemAbstract<FieldElement>>(
...
             &Boolean::Constant(true),
         )?;
+
+    }
+
+    if sc_prev_hash_g.is_some() {
+        let prev_hash = sc_prev_hash_g.unwrap();
+        let expected_prev_hash = expected_sc_prev_hash_g.unwrap();
+
+        for i in 0..remaining_constraints {
```
Comment (DanieleDiBenedetto) [inline src/mc_test_circuits/cert.rs:96]: We must update the calculations above, otherwise it does never enter in this for loop

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]
So as I understood the logic zend side should be:

- If Sidechain v2, `prev_cert_data_hash ` is passed the correct value if it's not the first certificate, and if it's the first one you pass PHANTOM_CERT_DATA_HASH using the corresponding function exposed in the zendoo_mc.h `zendoo_get_phantom_cert_data_hash`  (let's update the comment there);
- If Sidechain v1 then you simply omit the public input

Is it right ?

Also, I would add test for all the cases here too (as it's easier to do modifications), in `sampleCalls.cpp`, in the python test in `mc_test/callFromMTFExample.py` and finally in the Rust tests

File: src/mc_test_circuits/cert.rs (hunk 1 above)
Comment (drgora) [reply src/mc_test_circuits/cert.rs]: Makes sense, thanks, commit 16b63d9187a6cffded5290597e8e23ecee799ee5 should fix the size of the circuit.

File: mc_test/mcTestCall.cpp [hunk 2]
```diff
@@ -193,6 +193,7 @@ void create_verify_test_cert_proof(std::string ps_type_raw, std::string cert_typ
         (path_char_t*)proof_path.c_str(),
         proof_path_len,
         num_constraints,
+        nullptr, // do not consider V2 SC here
```
[the remark this answers is not in the document]
Comment (drgora) [reply mc_test/mcTestCall.cpp:196]: We have an updated version of `mcTestCall` in the `zend` repository. That version handles the `prev_cert_hash` field as expected to implement the key rotation functionality.

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: mc_test/mcTestCall.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanieleDiBenedetto) [reply mc_test/mcTestCall.cpp:196]: Would be nice to have it also here, so we can test using the callFromMTFExample.py. Later we can think of unifying the mcTestCall.cpp file for both zend and mc-cryptolib (maybe let's open a related issue in one of the two repos)

**Comment by drgora** [comment]
> So as I understood the logic zend side should be:
> 
>     * If Sidechain v2, `prev_cert_data_hash ` is passed the correct value if it's not the first certificate, and if it's the first one you pass PHANTOM_CERT_DATA_HASH using the corresponding function exposed in the zendoo_mc.h `zendoo_get_phantom_cert_data_hash`  (let's update the comment there);
> 
>     * If Sidechain v1 then you simply omit the public input
> 
> 
> Is it right ?
> 

Yes that's right.

> Also, I would add test for all the cases here too (as it's easier to do modifications), in `sampleCalls.cpp`, in the python test in `mc_test/callFromMTFExample.py` and finally in the Rust tests

I added one Rust test that uses `prev_cert_data_hash` to create and verify a proof. We have python tests that use the updated version of mcTestCall in the zend repository. I can add more tests in `sampleCalls.cpp` once it's fixed.

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: examples/sampleCalls.cpp [hunk 3]
```diff
@@ -1252,7 +1391,7 @@ TEST_SUITE("ZendooBatchProofVerifier") {
                 false, constant, sc_id, epoch_number, quality, NULL, 0,
                 NULL, 0, end_cum_comm_tree_root, btr_fee, ft_min_amount, sc_pk,
                 (path_char_t*)proof_path.c_str(), proof_path.size(), NUM_CONSTRAINTS,
-                &ret_code, true, &SUPPORTED_SEGMENT_SIZE
+                NULL, &ret_code, true, &SUPPORTED_SEGMENT_SIZE
```
Comment (DanieleDiBenedetto) [inline examples/sampleCalls.cpp:1394]: It would be nice to add test cases in which prev_cert_hash is present also in the ZendooBatchVerifier

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: examples/sampleCalls.cpp (hunk 3 above)
Comment (drgora) [reply examples/sampleCalls.cpp]: Done with e9273b2fefa804357570c0b4209949d6a81e9d26

**Review by DanieleDiBenedetto — APPROVED** [review]

**Pull request merged**